meadow-client/test/Marlowe/BlocklyTests.purs

```diff
@@ -0,0 +1,80 @@
+import Blockly.Generator (Generator, getInputWithName, inputList, workspaceToCode)
+import Blockly.Headless (newBlock)
+import Blockly.Headless as Headless
+import Blockly.Types (BlocklyState)
+import Control.Alt ((<|>))
+import Control.Lazy (class Lazy)
+import Control.Monad.Gen (class MonadGen)
+import Control.Monad.Rec.Class (class MonadRec)
+import Data.Either (Either(..))
+import Data.Maybe (Maybe(..))
+import Effect (Effect)
+import Effect.Class (class MonadEffect, liftEffect)
+import Effect.Unsafe (unsafePerformEffect)
+import Marlowe.Blockly (blockDefinitions, buildGenerator, toBlockly)
+import Marlowe.Gen (genContract)
+import Marlowe.Parser as Parser
+import Marlowe.Types (Contract, Observation, Value)
+import Test.QuickCheck (class Testable, Result, (===))
+import Test.QuickCheck.Gen (Gen)
+import Test.Unit (Test, TestSuite, suite, test)
+import Test.Unit.QuickCheck (quickCheck)
+import Text.Parsing.Parser (ParseError, runParser)
+import Text.Parsing.Parser.Basic (parens)
+
+all :: TestSuite
+all =
+  suite "Marlowe.Blockly" do
+    test "c2b2c" $ quickCheckGen c2b2c
+
+quickCheckGen :: forall prop. Testable prop => Gen prop -> Test
+quickCheckGen = quickCheck
+
+mkTestState :: forall m. MonadEffect m => m {blocklyState :: BlocklyState, generator :: Generator}
+mkTestState = do
+  blocklyState <- liftEffect $ Headless.createBlocklyInstance
+  liftEffect $ Blockly.addBlockTypes blocklyState blockDefinitions
+  liftEffect $ Headless.initializeWorkspace blocklyState
+  generator <- liftEffect $ buildGenerator blocklyState
+  pure {blocklyState: blocklyState, generator: generator}
+
+c2b2c :: forall m. MonadGen m => MonadRec m => Lazy (m Value) => Lazy (m Observation) => Lazy (m Contract) => m Result
+c2b2c = do
+  contract <- genContract
+  -- Unfortunately quickcheck runs the concrete Gen monad and it would need to be re-written to use MonadGen
+  -- https://github.com/purescript/purescript-quickcheck/blob/v5.0.0/src/Test/QuickCheck.purs#L97
+  -- I made the executive decision that it's not worth my time to do it in this specific case
+  -- I have created https://github.com/purescript/purescript-quickcheck/issues/102
+  let
+    result = unsafePerformEffect $ runContract contract
+  pure (result === Right contract)
+
+runContract :: Contract -> Effect (Either ParseError Contract)
+runContract contract = do
+  state <- liftEffect mkTestState
+  liftEffect $ buildBlocks state.blocklyState contract
+  code <- liftEffect $ workspaceToCode state.blocklyState state.generator
+  let
+    result = runParser code (parens Parser.contract <|> Parser.contract)
+  pure result
+
+buildBlocks :: BlocklyState -> Contract -> Effect Unit
+buildBlocks bs contract = do
+  let
+    mContract = getBlockById bs.workspace "root_contract"
+  rootBlock <- case mContract of
+    Nothing -> newBlock bs.workspace "contract"
+    Just block -> pure block
+  let
+    inputs = inputList rootBlock
+
+    mInput = getInputWithName inputs "contract"
+  case mInput of
```

Comment:
Slamdata often use this trick on `Maybe` types where they do:

```
for_ mInput $ \i -> 
  toBlockly newBlock bs.workspace i contract
```

(Where `for_` is just a flipped `traverse_`.)

I'm on the fence about it. What do you think?

Reply:
yeah, it looks good